### PR TITLE
CC Libraries - improvements and bug fixes

### PR DESCRIPTION
**Pull request by shaoshing**
## Improvements

* Make sure to update asset's preview image when it is edited outside of DS (e.g. Ai, or win Ps)
* Remove the last "updateDocument" in libraries actions.
* Prevent long asset title from scrolling horizontally when scrolling the asset list.

## Resolved issues

#1860 CC Lib: adding a vector smart object imports into library incorrectly (hopefully)
#2230 Positions of icons are off from clickable regions in Libraries panel header
#2201 Stale empty layer remains after placing cloud-linked smart object in new artboard

### CODE REVIEW

File: src/js/actions/documents.js [hunk 1]
```diff
@@ -605,7 +605,7 @@ define(function (require, exports) {
     };
     selectDocument.reads = [locks.JS_TOOL];
     selectDocument.writes = [locks.JS_APP];
-    selectDocument.transfers = [layerActions.resetLinkedLayers, historyActions.queryCurrentHistory,
+    selectDocument.transfers = ["layers.resetLinkedLayers", historyActions.queryCurrentHistory,
```
Comment (shaoshing) [inline src/js/actions/documents.js]: Change to string type action name to resolve circular dependencies between the layers and documents action.

File: src/js/models/layerstructure.js [hunk 2]
```diff
@@ -822,10 +822,13 @@ define(function (require, exports, module) {
...
-                // otherwise, replace the selected layer
-                var selectedLayers = document.layers.selected;
-                replaceLayer = selectedLayers && selectedLayers.size === 1 && selectedLayers.first();
+            } else if (this.selected && this.selected.size === 1) {
+                // otherwise, replace the selected layer if it has the same index with the new layer.
+                var selectedLayer = this.selected.first(),
+                    selectedLayerIndex = this.indexOf(selectedLayer),
+                    newLayerIndex = descriptors[0].itemIndex;
+                    
+                replaceLayer = selectedLayerIndex === newLayerIndex && selectedLayer;
```
Comment (shaoshing) [inline src/js/models/layerstructure.js]: turns out that PS does not replace the single-empty-selected layer after placing an embedded layer, so we need to check the new layer's index to determine whether we should do a replacement.

File: src/js/jsx/sections/libraries/Library.jsx [hunk 3]
```diff
@@ -70,6 +71,18 @@ define(function (require, exports, module) {
...
+        
+        shouldComponentUpdate: function (nextProps, nextState) {
+            // Library's modified time reflects itself and its elements, so there is no need to check its element's 
+            // modified time.
+            if (this._libraryLastModified === nextProps.library.modified &&
+                _.isEqual(this.state, nextState)) {
+                return false;
+            }
+            
+            this._libraryLastModified = nextProps.library.modified;
```
Comment (volfied) [inline src/js/jsx/sections/libraries/Library.jsx]: This update should be done in componentDidUpdate. It's not React-like to write into things in SCU calls.

File: src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx [hunk 4]
```diff
@@ -43,30 +43,60 @@ define(function (require, exports, module) {
         
         getInitialState: function () {
             return {
-                loading: true,
-                renditionPath: null
+                loading: false,
+                renditionPath: null,
+                lastModified: null
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx]: shouldn't this be `this.props.element.modified` ?

Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx]: I see the null helps with `_updateRenditionPath` starting on init... not really happy with the whole logic here... but oh well!

File: src/js/models/layerstructure.js (hunk 2 above)
Comment (volfied) [inline src/js/models/layerstructure.js]: This is a really hairy function... one of our weakest points. Did you test this with other replace layer interactions (like replacing a background layer, NOT replacing a background layer etc.)

**Comment by volfied** [comment]
Got couple questions @shaoshing

File: src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx (hunk 4 above)
Comment (shaoshing) [inline src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx]: I moved `lastModified` out of the state as it is not related to component updates, also made the code more clear by moving the `lastModified` assignment to the `updateRenditionPath` function.

File: src/js/models/layerstructure.js (hunk 2 above)
Comment (shaoshing) [inline src/js/models/layerstructure.js]: The change here depends on the descriptor attribute `itemIndex` from PS, so it should be reliable to do a replacement when the new layer has the same index with the background layer ? Here are some scenarios I tested and passed:

* select bg layer, draw a rectangular, bg layer replaced
* selecte multiple layers (including bg layer), draw a rectangular, bg layer NOT replaced
* select bg layer, drag asset (except for Ai asset, which will create a new layer) from libraries, bg layer replaced
* deselect bg layer, draw a rectangular, bg layer NOT replaced

Please let me know if I miss anything here.

File: src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx [hunk 5]
```diff
@@ -49,24 +49,54 @@ define(function (require, exports, module) {
...
-            var element = this.props.element;
+            this._updateRenditionPath();
+        },
+        
+        componentWillUpdate: function () {
+            this._updateRenditionPath();
+        },
+        
+        shouldComponentUpdate: function (nextProps, nextState) {
+            return this._lastModified !== nextProps.element.modified ||
```
Comment (volfied) [inline src/js/jsx/sections/libraries/assets/AssetPreviewImage.jsx]: `_lastModified` is not declared anywhere, can you declare it with a JSDoc at the top of the component?

**Comment by volfied** [comment]
One more, @shaoshing

**Comment by shaoshing** [comment]
Thank you for reviewing @volfied . I moved the `_lastModified` variable to a more significant location and added doc. Please check it out.

**Comment by volfied** [comment]
:+1: nice work

**Pull request merged**